src/core/qgsgdalutils.cpp

```diff
@@ -120,6 +186,132 @@ void QgsGdalUtils::resampleSingleBandRaster( GDALDatasetH hSrcDS, GDALDatasetH h
   GDALDestroyGenImgProjTransformer( psWarpOptions->pTransformerArg );
 }
 
+gdal::dataset_unique_ptr QgsGdalUtils::resampleImage( const QImage &image, const QgsRectangle &extent, QSize outputSize, GDALResampleAlg resampleAlg )
+{
+  gdal::dataset_unique_ptr srcDS = QgsGdalUtils::imageToMemoryDataset( image, extent, QgsCoordinateReferenceSystem() );
+  if ( !srcDS )
+    return nullptr;
+
+  gdal::dataset_unique_ptr destDS = QgsGdalUtils::createMultiBandMemoryDataset( GDT_Float32, 4, extent, outputSize.width(), outputSize.height(), QgsCoordinateReferenceSystem() );
+  if ( !destDS )
+    return nullptr;
+
+  gdal::warp_options_unique_ptr psWarpOptions( GDALCreateWarpOptions() );
+  psWarpOptions->hSrcDS = srcDS.get();
+  psWarpOptions->hDstDS = destDS.get();
+
+  psWarpOptions->nBandCount = 4;
+  psWarpOptions->panSrcBands = ( int * ) CPLMalloc( sizeof( int ) * 4 );
+  psWarpOptions->panDstBands = ( int * ) CPLMalloc( sizeof( int ) * 4 );
+  psWarpOptions->panSrcBands[0] = 1;
+  psWarpOptions->panSrcBands[1] = 2;
+  psWarpOptions->panSrcBands[2] = 3;
+  psWarpOptions->panSrcBands[3] = 4;
+  psWarpOptions->panDstBands[0] = 1;
+  psWarpOptions->panDstBands[1] = 2;
+  psWarpOptions->panDstBands[2] = 3;
+  psWarpOptions->panDstBands[3] = 4;
+
+  psWarpOptions->eResampleAlg = resampleAlg;
+
+  // Establish reprojection transformer.
+  psWarpOptions->pTransformerArg =
+    GDALCreateGenImgProjTransformer( srcDS.get(), GDALGetProjectionRef( srcDS.get() ),
+                                     destDS.get(), GDALGetProjectionRef( destDS.get() ),
+                                     FALSE, 0.0, 1 );
+  psWarpOptions->pfnTransformer = GDALGenImgProjTransform;
+
+  // Initialize and execute the warp operation.
+  GDALWarpOperation oOperation;
+  oOperation.Initialize( psWarpOptions.get() );
+
+  oOperation.ChunkAndWarpImage( 0, 0, GDALGetRasterXSize( destDS.get() ), GDALGetRasterYSize( destDS.get() ) );
+
+  GDALDestroyGenImgProjTransformer( psWarpOptions->pTransformerArg );
+  return destDS;
+}
+
+QImage QgsGdalUtils::resampleImage( const QImage &image, QSize outputSize, GDALResampleAlg resampleAlg )
+{
+  gdal::dataset_unique_ptr srcDS = resampleImage( image, QgsRectangle( 0, 0, image.width(), image.height() ), outputSize, resampleAlg );
+  if ( !srcDS )
+    return QImage();
+
+  QImage res( outputSize, image.format() );
+  QRgb *rgb = reinterpret_cast<QRgb *>( res.bits() );
+
+  GDALRasterBandH rBand = GDALGetRasterBand( srcDS.get(), 1 );
+  GDALRasterBandH gBand = GDALGetRasterBand( srcDS.get(), 2 );
+  GDALRasterBandH bBand = GDALGetRasterBand( srcDS.get(), 3 );
+  GDALRasterBandH aBand = GDALGetRasterBand( srcDS.get(), 4 );
+
+  const qgssize count = static_cast< qgssize >( outputSize.width() ) * outputSize.height();
+
+  float *rData = ( float * ) CPLMalloc( count * sizeof( float ) );
+  CPLErr err = GDALRasterIO( rBand, GF_Read, 0, 0, outputSize.width(), outputSize.height(),
```

Comment:
same as above (in reversed direction): you could likely GDALRasterIO() directly into QRgb* rgb